File: eda/algorhitms/Python/selectionSort/selectionSort.py
def selectionSort(array, start, end):
    for i in range(start, end + 1):
        indexSmallest = getIndexSmallest(array, start + i, end)
        swap(array, indexSmallest, i)


def getIndexSmallest(array, start, end):
    indexSmallest = start
    for i in range(start, end + 1):
        if array[i] < array[indexSmallest]:
            indexSmallest = i
    return indexSmallest


def swap(array, id1, id2):
    temp = array[id1]
    array[id1] = array[id2]
    array[id2] = temp


array = [9, 3, 7, 99, 5, 2]
selectionSort(array, 0, len(array) - 1)
print(array)

File: leda/R01-01/R01-01-Rot-SimpleSorting-Bidirectional-Bubble-environment/src/main/java/sorting/simpleSorting/SelectionSort.java
package sorting.simpleSorting;

import sorting.AbstractSorting;
import util.Util;

/**
 * The selection sort algorithm chooses the smallest element from the array and
 * puts it in the first position. Then chooses the second smallest element and
 * stores it in the second position, and so on until the array is sorted.
 */
public class SelectionSort<T extends Comparable<T>> extends AbstractSorting<T> {

    private void selectionSort(T[] array, int start, int end) {
        if(array.length == 0 || array.length == 1) {
            return;
        }
        for (int i = start; i < end; i++) {
            int indexSmallest = findIndexSmallest(array, i, end);

            boolean smallestIsInRightPosition = start == indexSmallest;
            if (!smallestIsInRightPosition) {
                Util.swap(array, i, indexSmallest);
            }
        }
    }

    private int findIndexSmallest(T[] array, int start, int end) {
        int indexSmallest = start;
        for (int i = start; i <= end; i++) {
            boolean isCurrentSmallerThanNext = array[i].compareTo(array[indexSmallest]) < 0;
            if (isCurrentSmallerThanNext) {
                indexSmallest = i;
            }
        }
        return indexSmallest;
    }

    @Override
    public void sort(T[] array, int leftIndex, int rightIndex) {
        selectionSort(array, leftIndex, rightIndex);
    }
}
